### PR TITLE
Refactor SaveOrderConfigPanel

### PR DESCRIPTION
**Pull request by calixtus**
A messy job, that was on my ToDo list for some time now.
I gave the SaveOrderConfigPanel an overhaul to make it a little bit more generic and to be able to work with more than three SortCriteria.

Before:
grafik

After:
grafik

One Problem remaining is that the preferences do only allow three SortCriteria for Export at all with the current JabRefPreferences implementation. Changing that would mean I would have to add PreferenceMigrations OR i just allow three SortCriteria for now in the Export prefs tab.

Yet some very little issues left, i hope i can fix them the next days.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Good! While you are at it, can you please take a look at https://discourse.jabref.org/t/sort-order-for-export/2598/12

**Comment by calixtus** [comment]
Hmm... I think I unintentionally stumbled over a bug in the tests.
The 'expected' value of in the test bib file is
image
while the test states that
image

**Comment by calixtus** [comment]
> Good! While you are at it, can you please take a look at https://discourse.jabref.org/t/sort-order-for-export/2598/12

I noticed, that when i change the number of criterias and save it, JabRef complains about the file being saved by another program. Could have something to do with it?

I can take a deeper look into this, but I'm somewhat careful, because I don't want this PR to explode beyond dealing with the visual representation. Let's see where this is going.

**Comment by calixtus** [comment]
DevCall decision:
- limit in preferences export to 3 criteria
- debug https://discourse.jabref.org/t/sort-order-for-export/2598/12 in a followup pr

**Review by koppor — APPROVED** [review]

**Review by koppor — APPROVED** [review]

**Comment by koppor** [comment]
Fixes https://github.com/koppor/jabref/issues/510

File: src/main/java/org/jabref/model/metadata/SaveOrderConfig.java [hunk 1]
```diff
@@ -40,15 +66,21 @@ private SaveOrderConfig(List<String> data) {
...
+            } else {
+                LOGGER.warn("Could not parse sort order: {}", data.get(0));
+                this.orderType = OrderType.ORIGINAL;
+                return;
+            }
         }
 
         for (int index = 1; index < data.size(); index = index + 2) {
-            sortCriteria.addLast(new SortCriterion(FieldFactory.parseField(data.get(index)), data.get(index + 1)));
+            sortCriteria.add(new SortCriterion(FieldFactory.parseField(data.get(index)), data.get(index + 1)));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/metadata/SaveOrderConfig.java:83]: Why did this change?

**Review by Siedlerchr — APPROVED** [review]

File: src/main/java/org/jabref/model/metadata/SaveOrderConfig.java (hunk 1 above)
Comment (calixtus) [reply src/main/java/org/jabref/model/metadata/SaveOrderConfig.java:83]: @koppor suggested, that i should switch from LinkedLists to ArrayLists. ArrayLists do not provide a method addLast, but only add, which should add the elements at the end of the list anyway.

**Pull request merged**